### PR TITLE
Add missing discriminators to all remaining unions with scalars

### PR DESCRIPTION
**Pull request by donbox**
KeyValuePairProto and Dimension now have proto3 discriminators.

### CODE REVIEW

**Comment by donbox** [comment]
Folks, this is my first PR since we moved to onnx.proto.in, hence the build break.

File: onnx/onnx.proto.in [hunk 1]
```diff
@@ -345,6 +348,15 @@ message SparseTensorProto {
...
   message KeyValuePairProto {
+    enum KeyType {
+      UNDEFINED = 0;
+      STRING = 1;
+      INT32 = 2;
+      INT64 = 3;
+      UINT64 = 4;
+    }
+    // This field MUST be present for this version of the IR.
+    optional KeyType type = 5;
```
Comment (linkerzhang) [inline onnx/onnx.proto.in]: We don't need the enum in this case, as we have already used oneof there. There will be a "enum" generated with oneof usage.

File: onnx/onnx.proto.in [hunk 2]
```diff
@@ -428,7 +440,15 @@ message TypeProto {
   // or a symbolic variable. A symbolic variable represents an unknown
   // dimension.
   message TensorShapeProto {
+    message DimensionType {
+      UNDEFINED = 0;
+      VALUE = 1;
+      PARAM = 2;
+    }
+
```
Comment (linkerzhang) [inline onnx/onnx.proto.in]: Same as above please.

**Comment by prasanthpul** [comment]
I'm not sure we need this PR. KeyValuePairProto and Dimension already use oneof

**Comment by donbox** [comment]
I believe that we need it even for oneof cases given the HasField issues in Proto3

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by gramalingam** [comment]
I think "oneof key {...}" will generate a corresponding enum in Proto3. I think the problematic case is "optional" fields of scalar-type (which are not in an enum) ... these will generate "HasField" in Proto2 but not in Proto3.

**Comment by gramalingam** [comment]
Typo correction: I meant "not in an oneof" above (and not "not in an enum").

**Comment by prasanthpul** [comment]
https://developers.google.com/protocol-buffers/docs/reference/python-generated#oneof 

You should be able to use HasField on oneof in proto3

**Comment by jywumsft** [comment]
For C# proto3 does not generate *any* HasField() methods. 

submessages will be null if they were not set. 

For oneof, an enum will be generated so it can be used to test presence. e.g. 
TypeProto.Types.TensorShapeProto.Types.Dimension.ValueOneofCase

For primitive types, they get filled in with default values if not explicitly set,
thus we need the discriminator (i.e. what was done for AttributeProto)

**Comment by prasanthpul** [comment]
Thanks for clarifying @jywumsft! Looks like we don't need this PR. @donbox do you still feel otherwise?

**Pull request closed without merging**